brainiak/eventseg/event.py

```diff
@@ -0,0 +1,356 @@
+        self.p_end = np.zeros((1, self.n_events + 1))
+        self.p_end[0, -2] = 1
+
+        self.classes_ = np.arange(self.n_events)
+
+        # Initialize variables for fitting
+        log_gamma = []
+        for i in range(n_train):
+            log_gamma.append(np.zeros((X[i].shape[1],
+                                       self.n_events)))
+        step = 1
+        best_ll = float("-inf")
+        self.ll_ = np.empty((0, n_train))
+        while step <= self.n_iter:
+            iteration_var = self.step_var(step)
+
+            # Based on the current segmentation, compute the mean pattern
+            # for each event
+            seg_prob = [np.exp(lg) / np.sum(np.exp(lg), axis=0)
+                        for lg in log_gamma]
+            mean_pat = np.empty((n_train, n_dim, self.n_events))
+            for i in range(n_train):
+                mean_pat[i, :, :] = X[i].dot(seg_prob[i])
+            mean_pat = np.mean(mean_pat, axis=0)
+
+            # Based on the current mean patterns, compute the event
+            # segmentation
+            self.ll_ = np.append(self.ll_, np.empty((1, n_train)), axis=0)
+            for i in range(n_train):
+                logprob = self._logprob_obs(X[i],
+                                            mean_pat, iteration_var)
+                log_gamma[i], self.ll_[-1, i] = self._forward_backward(logprob)
+
+            # If log-likelihood has started decreasing, undo last step and stop
+            if np.mean(self.ll_[-1, :]) < best_ll:
+                self.ll_ = self.ll_[:-1, :]
+                break
+
+            self.segments_ = [np.exp(lg) for lg in log_gamma]
+            self.event_var_ = iteration_var
+            self.event_pat_ = mean_pat
+            best_ll = np.mean(self.ll_[-1, :])
+            logger.debug("Fitting step %d, LL=%f", step, best_ll)
+
+            step += 1
+
+        return self
+
+    def _logprob_obs(self, data, mean_pat, var):
+        """Compute the log probability of each observed timepoint being
+        generated by the Gaussian distribution for each event pattern
+
+        Parameters
+        ----------
+        data: voxel by time ndarray
+            fMRI data on which to compute log probabilities
+
+        mean_pat: voxel by event ndarray
+            Centers of the Gaussians for each event
+
+        var: float or 1D array of length equal to the number of events
+            Variance of the event Gaussians. If scalar, all events are
+            assumed to have the same variance
+
+        Returns
+        -------
+        logprob : time by event ndarray
+            Log probability of each timepoint under each event Gaussian
+        """
+
+        n_vox = data.shape[0]
+        t = data.shape[1]
+
+        # z-score both data and mean patterns in space, so that Gaussians
+        # are measuring Pearson correlations and are insensitive to overall
+        # activity changes
+        data = stats.zscore(data, axis=0, ddof=1)
```

Comment:
I suspect that this line might run the same risk: it will z-score the original data passed to the fit() function, unless the original data is passed as a copy.